### DISCUSSION TITLE
Filter with inner joins but still get the complete inner data

### DISCUSSION THREAD

**Discussion by jpcafe10** [supabase/supabase#6587] [Questions] [answered]
Hi all,

Given the following query where `modelA` can have multiple tags associated to it

```js
supabase.from('modelA').select('id, tags!inner(id,name)').in('tags.id', ['some tag'])
```

Is it possible to filter `modelA` that includes `tag.id` but still get all the tags associated to it in the response?

I'm seeing the correct top level filtering for `modelA` but it only includes the tags that were a part of the filtering.

**Comment by steve-chavez** [comment]
> Is it possible to filter modelA that includes tag.id but still get all the tags associated to it in the response?

Maybe try the [is](https://supabase.com/docs/reference/javascript/is) operator combined with [not](https://supabase.com/docs/reference/javascript/not).

If the above doesn't help, perhaps you can show sample data of what you currently get and what your desired response would be?

**Reply by jpcafe10** [reply]
```js
// Tags has many-to-one relation, myModel can have many tags

const dataset = [
  {
    id: 1,
    tags: [
      { id: 1, name: 'tag-1' },
      { id: 2, name: 'tag-2' },
    ],
  },
  {
    id: 2,
    tags: [{ id: 3, name: 'tag-3' }],
  },
];

supabase.from('myModel').select('id, tags!inner(id,name)').in('tags.id', [1]);

const currentResult = [
  {
    id: 1,
    tags: [{ id: 1, name: 'tag-1' }],
  },
];

```

But what i would like to achieve is

```js
const currentResult = [
  {
    id: 1,
    tags: [
      { id: 1, name: 'tag-1' },
      { id: 2, name: 'tag-2' },
    ],
  },
];



```

**Reply by steve-chavez** [reply]
Ah, I see. That's not how the inner join works unfortunately. It will only return rows common to both the parent(modelA) and child(tags).

Would it be possible for you to have tags as a `text[]`(or `jsonb[]`) array instead? Then your query could be solved with just a [contains](https://supabase.com/docs/reference/javascript/contains).

Otherwise, we'd be looking at doing an aggregation(`array_agg` or `json_agg`) on a function or view of `tags` and then filter on those; a bit more complicated and likely with worse perf too.

**Reply by jpcafe10** [reply]
That's okay, it's for a personal project anyways 😄 
The `text[]` or `json[]` option would lose the one-to-many dependency, correct? ie. would be a plain array of things.

**Reply by steve-chavez** [reply]
> The text[] or json[] option would lose the one-to-many dependency, correct? ie. would be a plain array of things.

Yes, correct.

**Comment by jpcafe10** [comment, marked as answer]
I found a solution for this.

You can add a temp alias to do your inner join filtering while still selecting the data you require. For ex:

```js

supabase.from('myModel').select('id, tags(id,name), temp_tags:tags!inner(id,name)').in('temp_tags.id', [1]);
```

This way you apply your inner filtering and still get the complete list of children.

**Reply by steve-chavez** [reply]
That is awesome. It's basically a LEFT JOIN(tags) plus a WHERE EXISTS(temp_tags) in SQL.

I assume you don't really want the `temp_tags` output. On the upcoming version of the REST API(PostgREST), you'll be able to do an `exists` filter where it won't be needed to include `temp_tags` in `select`.

**Reply by ge0dem0** [reply]
@jca41 you are awesome! I was looking for this solution for ages. Thank you so much for sharing!! 🙏

**Reply by steve-chavez** [reply]
> On the upcoming version of the REST API(PostgREST), you'll be able to do an exists filter where it won't be needed to include temp_tags in select.

FYI, nowadays you can leave the `temp_tags` empty (https://docs.postgrest.org/en/v12/references/api/resource_embedding.html#empty-embed):

```js
supabase.from('myModel').select('id, tags(id,name), temp_tags:tags!inner()').in('temp_tags.id', [1]);
```

**Reply by christophemenager** [reply]
Thanks a lot @steve-chavez for the updated version! 🙏
